x-pack/legacy/plugins/lens/public/editor_frame_plugin/editor_frame/save.test.ts

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { save } from './save';
+import { Action } from './state_management';
+
+describe('save editor frame state', () => {
+  it('dispatches saved status actions before and after saving', async () => {
+    let saved = false;
+
+    const dispatch = jest.fn((action: Action) => {
+      if (action.type === 'SAVING' && saved) {
+        throw new Error('Saving was called after save');
+      }
+      if (action.type === 'SAVED' && !saved) {
+        throw new Error('Saved was called before save');
+      }
+    });
+
+    await save({
+      dispatch,
+      redirectTo: jest.fn(),
+      state: {
+        title: 'aaa',
+        datasource: { activeId: '1', isLoading: false, state: {} },
+        saving: false,
+        visualization: { activeId: '2', state: {} },
+      },
+      store: {
+        async save() {
+          saved = true;
+          return { id: 'foo' };
+        },
+      },
+    });
+
+    expect(dispatch).toHaveBeenCalledWith({ type: 'SAVING' });
+    expect(dispatch).toHaveBeenCalledWith({ type: 'SAVED' });
+  });
+
+  it('transforms from internal state to persisted doc format', async () => {
+    const store = {
+      save: jest.fn(async () => ({ id: 'bar' })),
+    };
+    await save({
+      store,
+      datasource: {
+        getPersistableState(state) {
+          return {
+            stuff: `${state}_datsource_persisted`,
+          };
+        },
+      },
+      dispatch: jest.fn(),
+      redirectTo: jest.fn(),
+      state: {
+        title: 'bbb',
+        datasource: { activeId: '1', isLoading: false, state: '2' },
+        saving: false,
+        visualization: { activeId: '3', state: '4' },
+      },
+      visualization: {
+        getPersistableState(state) {
+          return {
+            things: `${state}_vis_persisted`,
+          };
+        },
+      },
+    });
+
+    expect(store.save).toHaveBeenCalledWith({
+      datasourceType: '1',
+      id: undefined,
+      state: {
+        datasource: { stuff: '2_datsource_persisted' },
+        visualization: { things: '4_vis_persisted' },
+      },
+      title: 'bbb',
+      type: 'lens',
+      visualizationType: '3',
+    });
+  });
+
+  it('redirects to the edit screen if the id changes', async () => {
+    const id = `${Math.random()}`;
```

Comment:
I can change it. I've seen tests like this pass because they redirected to a specific URL, for example, where randomizing the test would have caught it.